src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -317,39 +343,68 @@ private SinkRecord handleNativeRecord(SinkRecord record, boolean isKey) {
         record.headers());
   }
 
-  /** Invokes insertRows API using the offsets buffered for this partition. */
-  public InsertValidationResponse insertBufferedRows() {
-    PartitionBuffer intermediateBuffer = null;
-    bufferLock.lock();
-    try {
-      intermediateBuffer = streamingBuffer;
-      streamingBuffer = new StreamingBuffer();
-    } finally {
-      // release lock
-      bufferLock.unlock();
+  // --------------- BUFFER FLUSHING LOGIC --------------- //
+  /**
+   * If difference between current time and previous flush time is more than threshold, insert the
+   * buffered Rows.
+   *
+   * <p>Note: We acquire buffer lock since we copy the buffer.
+   *
+   * <p>Threshold is config parameter: {@link
+   * com.snowflake.kafka.connector.SnowflakeSinkConnectorConfig#BUFFER_FLUSH_TIME_SEC}
+   *
+   * <p>Previous flush time here means last time we called insertRows API with rows present in
+   */
+  protected void insertBufferedRowsIfFlushTimeThresholdReached() {
+    if (this.streamingBufferThreshold.isFlushTimeBased(this.previousFlushTimeStampMs)) {
+      LOGGER.debug(
+          "Time based flush for channel:{}, CurrentTimeMs:{}, previousFlushTimeMs:{},"
+              + " bufferThresholdMs:{}",
+          this.getChannelName(),
+          System.currentTimeMillis(),
+          this.previousFlushTimeStampMs,
+          this.streamingBufferThreshold.getFlushTimeThresholdSeconds());
+      StreamingBuffer copiedStreamingBuffer;
+      bufferLock.lock();
+      try {
+        copiedStreamingBuffer = this.streamingBuffer;
+        this.streamingBuffer = new StreamingBuffer();
+      } finally {
+        bufferLock.unlock();
+      }
+      insertBufferedRows(copiedStreamingBuffer);
```

Comment:
Not sure if you need to check if `copiedStreamingBuffer` is null (at least you did it in line 254